k2/csrc/weights.cc

```diff
@@ -68,11 +69,67 @@ WfsaWithFbWeights::WfsaWithFbWeights(const Fsa &fsa, const float *arc_weights,
     : fsa(fsa), arc_weights(arc_weights), weight_type(t) {
   if (IsEmpty(fsa)) return;
   CheckInput(fsa, arc_weights);
+  ComputeForwardWeights();
+  ComputeBackardWeights();
+}
+
+// Mohri, M. 2002. Semiring framework and algorithms for shortest-distance
+// problems, Journal of Automata, Languages and Combinatorics 7(3): 321-350,
+// 2002.
+void WfsaWithFbWeights::ComputeForwardWeights() {
+  auto num_states = fsa.NumStates();
+  forward_state_weights = std::unique_ptr<double[]>(new double[num_states]);
+  std::fill_n(forward_state_weights.get(), num_states, kDoubleNegativeInfinity);
+
+  auto add_weights_func = AddWeights();
+  const auto &arcs = fsa.arcs;
+  forward_state_weights[0] = 0;
+  for (std::size_t i = 0; i != arcs.size(); ++i) {
+    const auto &arc = arcs[i];
+    DCHECK_GE(arc.dest_state, arc.src_state);
+    auto src_weight = forward_state_weights[arc.src_state];
+    auto &dest_weight = forward_state_weights[arc.dest_state];
+
+    double r = src_weight + arc_weights[i];
+    dest_weight = add_weights_func(dest_weight, r);
+  }
+}
 
+void WfsaWithFbWeights::ComputeBackardWeights() {
   auto num_states = fsa.NumStates();
-  forward_state_weights = std::unique_ptr<double[]>(new double[num_states]());
-  backward_state_weights = std::unique_ptr<double[]>(new double[num_states]());
-  // TODO(haowen): compute `forward/backward_state_weights`, also check
-  // `IsTopSorted(fsa)`
+  backward_state_weights = std::unique_ptr<double[]>(new double[num_states]);
+  std::fill_n(backward_state_weights.get(), num_states,
+              kDoubleNegativeInfinity);
+
+  auto add_weights_func = AddWeights();
+  const auto &arcs = fsa.arcs;
+  backward_state_weights[fsa.FinalState()] = 0;
+  for (auto i = static_cast<int32_t>(arcs.size()) - 1; i >= 0; --i) {
+    const auto &arc = arcs[i];
+    DCHECK_GE(arc.dest_state, arc.src_state);
+    auto &src_weight = backward_state_weights[arc.src_state];
+    auto dest_weight = backward_state_weights[arc.dest_state];
+
+    double r = dest_weight + arc_weights[i];
+    src_weight = add_weights_func(src_weight, r);
+  }
 }
+
+WfsaWithFbWeights::AddWeightsFun WfsaWithFbWeights::AddWeights() const {
+  switch (weight_type) {
+    case kMaxWeight:
+      return [](double w1, double w2) { return std::max(w1, w2); };
+      break;
+    case kLogSumWeight:
+      return [](double w1, double w2) { return LogAdd(w1, w2); };
+      break;
+    default:
+      // just return a value to compress compiler warnings,
+      // the program will be terminated after `LOG(FATAL)`
+      return [](double w1, double w2) { return LogAdd(w1, w2); };
```

Comment:
It should be placed after LOG(FATAL), otherwise, LOG(FATAL) is never reached.

Reply:
Oh, sorry, .